idx/index/types.go

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package index
+
+import (
+	"github.com/m3db/m3ninx/idx/doc"
+)
+
+// Index is a collection of searchable documents.
+type Index interface {
+	Writer
+
+	// Snapshot returns a point-in-time snapshot of the index that provides
+	// a consistent view for searches.
+	Snapshot() (Snapshot, error)
+}
+
+// Writer is the interface for adding documents to an index.
+type Writer interface {
+	// Open initializes the index for use.
+	Open() error
+
+	// Insert inserts the given document into the index. The document is
+	// guaranteed to be searchable once the Insert method returns.
+	Insert(d doc.Document) error
+
+	// Closes closes the index.
+	Close() error
+}
+
+// Snapshot is a point-in-time snapshot of an index. It must be closed to
+// release the internal readers.
+type Snapshot interface {
```

Comment:
I think you'll need a couple more methods here:
- Ability to go from DocID(s)/PostingsList -> Documents
- Ability to iterate known documents in the snapshot

Reply:
Hmm, I wonder if that wouldn't be better left entirely to the Readers? For example, the first method is provided by the `Docs` method on a `Reader`. There's no method for the second though it could be added. I'm curious why you feel there's a need to iterate over all the documents known to the `Reader`? Said another way, I'm just not aware of a use case that needs the ability to iterate over all documents instead of being to just search over them with a query.

Reply:
Per our offline chat, either Index/Segment will need the Iterator to allow for other sub-systems to interact with them. e.g.  The type responsible for persisting a segment/index to disk would use this.